docs/design/plugins.md

```diff
@@ -0,0 +1,164 @@
+# Plugins
+
+This file describes a plugin mechanism for Smith. A plugin can be used to transform
+dependencies and their outputs into specifications of objects to be created.
+
+## Motivation
+
+A typical example when a runtime transformation is needed is two `ServiceInstance`s, one producing inputs for the
+other. Say `b` depends on `a`. Quite often the shape of data `ServiceInstance` `a` produces does not match what
+`ServiceInstance` `b` expects.
+
+```yaml
+apiVersion: smith.atlassian.com/v1
+kind: Bundle
+metadata:
+  name: ab
+spec:
+  resources:
+
+  - name: a
+    spec:
+      apiVersion: servicecatalog.k8s.io/v1beta1
+      kind: ServiceInstance
+      metadata:
+        name: a
+      spec:
+        foo: bar
+
+  - name: a-binding
+    dependsOn:
+    - a
+    spec:
+      apiVersion: servicecatalog.k8s.io/v1beta1
+      kind: ServiceBinding
+      metadata:
+        name: a-binding
+      spec:
+        instanceRef:
+          name: "{{a#metadata.name}}"
+        secretName: a-binding-secret
+
+  - name: b
+    dependsOn:
+    - a
+    spec:
+      apiVersion: servicecatalog.k8s.io/v1beta1
+      kind: ServiceInstance
+      metadata:
+        name: b
+      spec:
+        # need data from a-binding-secret but transformed
+        # e.g. only keys which start with "FOO_" when full names of the keys are not known in advance.
+```
+
+## Specification
+
+Smith plugins are [Go plugins](https://golang.org/pkg/plugin/). They are eagerly loaded at Smith startup to detect
+issues early. Names of plugins to load are passed via command line.
+Each plugin publishes a `Process(smith_v1.Resource, map[smith_v1.ResourceName]Dependency) (TransformResult, error)`
+function.
+
+When Smith comes across a resource with `type: plugin` and `pluginName: foobar` it invokes
+the plugin `foobar`. To do that it inspects all the dependencies of the resource specified in `dependsOn` attribute
+and fetches their outputs to include in the invocation along with the dependencies themselves. Smith needs to
+recognize resource group/version/kinds to be able to fetch the output (if there is one).
```

Comment:
Does this extend to allowing smith bundle objects to depend on objects from outside the bundle or is this the preferred mechanism to allow that to happen?